GH-130798: Add type hints to pathlib.types

ap-- opened the pull request:
Hello everyone,

This PR addresses python/cpython#130798

~As requested, the type hints are kept compatible with Python 3.9 to simplify release in `pathlib-abc`.~

I also just noticed #131621 which I assumed will be merged soon, because it simplifies typing of `_ReadablePath.copy` and `_ReadablePath.copy_into`. Otherwise, for `str` targets the `_ReadablePath` subclass has to be a `_WritablePath` subclass too to make sense as a target.

Since, I haven't contributed type hinted code to cpython so far, it would be great if you could let me know if this PR should also ship tests for verifying the types somewhere.

Cheers,
Andreas 😃

ap-- committed:
Add typehints to pathlib.types

picnixz commented on Lib/pathlib/types.py:
```diff
@@ -240,45 +248,50 @@ class _ReadablePath(_JoinablePath):
 
     @abstractmethod
-    def iterdir(self):
+    def iterdir(self: _RP) -> Generator[_RP, None, None]:
```
We can return an `Iterator` here. It doesn't matter that we return a generator I think.

ap-- replied:
I'll make the change.

picnixz commented on Lib/pathlib/types.py:
```diff
@@ -240,45 +248,50 @@ class _ReadablePath(_JoinablePath):
         raise NotImplementedError
 
-    def glob(self, pattern, *, recurse_symlinks=True):
+    def glob(self: _RP, pattern: Union[_RP, str], *, recurse_symlinks: bool = True) -> Generator[_RP, None, None]:
```
Ditto. Also, I think you want to chop down the params list.

Also, the pattern does not need to be of the same type as `self`.

ap-- replied:
I switched to `Iterator`, but haven't made the pattern type more generic.

Are you suggesting to use `Union[_JoinablePath, str]` here?

picnixz commented on Lib/pathlib/types.py:
```diff
@@ -370,29 +388,29 @@ class _WritablePath(_JoinablePath):
 
     @abstractmethod
-    def symlink_to(self, target, target_is_directory=False):
+    def symlink_to(self: _WP, target: _WP, target_is_directory: bool = False) -> None:
```
Ditto with the second `_WP` not necessarily equal to the first one IMO.

ap-- replied:
I thought about this one, and I don't think symlinks between different Path implementations should be supported. I agree that somehow intuitively I would expect target to be either a ReadablePath or WriteablePath of the same path type, but since WritablePath does not inherit from ReadablePath, this seems to be the best compromise.

Do you have a suggestion how to change the target type?

picnixz replied:
The target can also be a PathLike object, which is not always a WritablePath. 
For instance, `WritableLocalPath` implements:

```py
def symlink_to(self, target, target_is_directory=False):
    os.symlink(target, self, target_is_directory)
```

and I think `os.symlink` accepts PathLike objects. So, to avoid violating LSP,
I think we should maybe just put PathLike?

ap-- replied:
It's possible I am missing some context here.

`PathLike` implies local paths. But many remote file systems don't have symlink functionality, but can potentially emulate it.

For example a redirect on http filesystems could be considered a symlink, and I could see how to implement that functionality for a filesystem like that.

To be inline with #131621 and the fact that `str` targets are not considered for `.copy` and `.copy_into`, it seems logical to me to require `target` to be of the same base type as `self`.

picnixz replied:
What I meant with `WritableLocalPath` is that `WritableLocalPath` inherits `WritablePath`. So, whatever we choose as an argument type for `symlink_to` should be compatible with what `WritableLocalPath` expects. Let's say that we put `target` with the same base type as self. Then the issue is that `WritableLocalPath` is not allowed to accept a string argument for `target`, as the signature wouldn't be compatible anymore.

So, ideally, we should have a larger type in the parent signature and children classes should reject types that do not fit their signatures. For instance:

```py
# parent:
def symlink_to(self, target: U | V, ...) -> None: ...

# child:
@overloads
def symlink_to(self, target: U, ...) -> NoReturn: ...
@overloads
def symlink_to(self, target: V, ...) -> None: ...
def symlink_to(self, target: U | V, ...) -> Any: ...
``` 

If we were to already restrict target to the same type as self, then we are also restricting the child interfaces, e.g., `WritableLocalPath`.

I don't know how Barney wants to deal with this or whether `WritableLocalPath` should accept strings. If it shouldn't, we can indeed keep what you wrote.

barneygale replied:
I think target should be a `str`, if that makes sense to folks?

User implementations might extend this to `str | Self`, and in `pathlib.Path` it would be `os.PathLike[str]` (which includes `str` itself)

barneygale replied:
(Agree that `os.PathLike` should **not** feature in any of these type hints.)

ap-- replied:
target is now `str`

picnixz reviewed (commented):
~~I'm worried about the covariant returns that are lacking but I don't remember if it's allowed to have a method with `def f(self: T_co) -> T_co: ...`~~

ap-- committed:
Iterator type for .glob and .iterdir

barneygale commented on Lib/pathlib/types.py:
```diff
@@ -72,38 +80,38 @@ class _JoinablePath(ABC):
 
     @abstractmethod
-    def with_segments(self, *pathsegments):
+    def with_segments(self: _JP, *pathsegments: Union[_JP, str]) -> _JP:
```
I think pathsegments should be `str` actually. It's not always possible to support joining two objects of the same type, e.g. two `ZipPath` objects with different underlying zip files.

barneygale replied:
This also applies to `joinpath()`, `__truediv__` and `__rtruediv__`

ap-- replied:
changed to `str`

barneygale commented on Lib/pathlib/types.py:
```diff
@@ -240,45 +248,50 @@ class _ReadablePath(_JoinablePath):
         raise NotImplementedError
 
-    def glob(self, pattern, *, recurse_symlinks=True):
+    def glob(self: _RP, pattern: Union[_RP, str], *, recurse_symlinks: bool = True) -> Iterator[_RP]:
```
`recurse_symlinks=False` is unsupported and always raises `NotImplementedError`. The argument exists for compatibility with `Path.glob()`, where the default is `False`.

Can we use typing to indicate that the value _must_ be `True`?

picnixz replied:
Yes: `recurse_symlinks: Literal[True] = True`

picnixz replied:
> The argument exists for compatibility with Path.glob(), where the default is False.

Technically, mypy will complain but this one can be "ignored" if we were to run mypy on Lib/*. But for anyone else, Literal[True] is sufficient (as this is what we expect; one could say that Path.glob() is the weird one)

barneygale replied:
It's a super annoying situation. Changing the default in `pathlib.Path.glob()` is probably too disruptive for users, but not recursing into symlinks is slower (pattern parts after `**` incur system calls, where's they're free if we recurse into symlinks), and the effect is even more pronounced when the filesystem is laggy (e.g. a virtual filesystem going over network, like many use cases for `ReadablePath`).

In theory we could deprecate `pathlib.Path.glob(recurse_symlinks=False)` once Python 3.12 falls out of security support, and force users to pass `recurse_symlinks=True` to avoid the deprecation warning. Then in a further ~5 years we could drop support for `recurse_symlinks=False`. But even that's not great - there will be a ~5 year period where everyone's calls to `glob()` will add `recurse_symlinks=True` effectively as boilerplate.

ap-- replied:
pattern is now `str` and recurse_symlinks is `Literal[True]`

ap-- committed:
Update types

- pathsegments are `str`: joinpath, __truediv__, __rtruediv__, with_segments
- symlink_to target is `str`
- glob recurse_symlinks is `Literal[True]`
- full_match and glob pattern is `str`

ap-- committed:
Merge remote-tracking branch 'origin/main' into pathlib-types-add-typehints

ap-- committed:
Update pathlib.types to use Self

ap-- committed:
Rely on python3.13 Generator defaults

picnixz commented on Lib/pathlib/types.py:
```diff
@@ -14,10 +14,15 @@
 from pathlib import PurePath, Path
-from typing import Optional, Protocol, runtime_checkable
+from typing import (
+    Any, BinaryIO, Callable, Generator, Iterator, Literal, Optional, Protocol, Self, Sequence, TypeVar,
```
For the stdlib, we can use `from collections.abc import Callable, Iterator, ...` etc.

ap-- replied:
I moved for all deprecated typing ABCs to collections.abc

picnixz commented on Lib/pathlib/types.py:
```diff
@@ -290,7 +300,12 @@ def glob(self, pattern, *, recurse_symlinks=True):
+        top_down: bool = True,
+        on_error: Optional[Callable[[Exception], None]] = None,
+        follow_symlinks: bool = False,
+    ) -> Generator[tuple[Self, list[str], list[str]]]:
```
Isn't it possible to simply use Iterable here as well?

ap-- replied:
I switched to Iterator

picnixz commented on Lib/pathlib/types.py:
```diff
@@ -14,10 +14,15 @@
 
-def _explode_path(path, split):
+
+def _explode_path(path: str, split: Callable[[str], tuple[str, str]]) -> tuple[str, list[str]]:
```
Does it fit on 80 chars?

barneygale replied:
(If you want to make `_explode_path()` accept a `parser: _PathParser` argument instead of `split`, that's fine, please go ahead)

ap-- replied:
I changed the signature to accept a parser

picnixz reviewed (commented):
For `write_*` methods, one could return int or int / None if you want children classes to only return None if it's too hard for them to know how many characters were written. WDYT @barneygale?

barneygale commented on Lib/pathlib/types.py:
```diff
@@ -322,21 +337,21 @@ def walk(self, top_down=True, on_error=None, follow_symlinks=False):
         raise NotImplementedError
 
-    def copy(self, target, **kwargs):
+    def copy(self, target: _WP, **kwargs: Any) -> _WP:
```
Is it possible to use a `ParamSpec`/similar here to show that the keyword arguments are passed to `target._copy_from()`?

ap-- replied:
I spent quite a bit of time now trying to get this to work, but I believe the way `_ReadablePath.copy` and `_WritablePath._copy_from` are coupled is currently not representable in the Python type system.

Be aware, I am not an expert in this, so all conclusions I draw here should be confirmed with someone who is.


Ideally the typechecker would just be able to infer if keyword arguments are not compatible with the signature of a detected subclass of `_WritablePath._copy_from` . But all my attempts failed so far. Basically the minimal scenario that needs to be supported is:

```python

class R:
    def copy[T: W](self, target: T, **kwargs) -> T:
        target._copy_from(self, **kwargs)
        return target

class W:
    def _copy_from(self, source: R, /, *, follow_symlinks: Literal[True] = True) -> None:
        return

```

1. In this example: "not typing the kwargs" (or typing them as Any) leads to missing incorrect keyword arguments: https://mypy-play.net/?mypy=latest&python=3.13&gist=7608e8dcd5e4b58dc2fc2c355bf6ed89

2. Trying to use TypedDicts to define the kwargs doesn't work either because the definitions would have to be provided to `_ReadablePath.copy` for each speficic target class: https://mypy-play.net/?mypy=latest&python=3.13&gist=12f4cbdd725073d8a152a90a941922cf

3. ParamSpec looks pretty promising but to be able to actually bind the parameters, you need to go via a protocol class, which prevents you from reusing the actual target type as return type of the `_ReadablePath.copy` method. Moreover ParamSpec can't be used without binding P.args to the function signature: https://mypy-play.net/?mypy=latest&python=3.13&gist=b47ea834a0f57a722b8124f76f3fb6d4

Ideally we'd be able to do something like:

```python
  class R:
      def copy[T: W, **P](self, target: T[P], **kwargs: P.kwargs) -> T[P]: ...
  class W[**P]:
      def _copy_from(self, source: R, **kwargs: P.kwargs) -> None: ...
  ```

But from reading through a bunch of docs and issues, I this seems impossible. I believe the correct related issue here should be this one python/typing#548 for support of higher-kinded typevars, and we'd need `kwargs_only` support for ParamSpec: python/typing#1524

My guess is that by redesigning the coupling between `_ReadablePath.copy` and `_WritablePath._copy_from` it should be possible to better represent this in the type system. But it's getting late here, and I think I'll revisit this with a fresh mind tomorrow.

TeamSpen210 replied:
[Intersection types](https://github.com/CarliJoy/intersection_examples) might also work, since then you could do the following:
```py
def copy[T: W, **P](self, target: T & SupportsCopyFrom[P], *args: P.args, **kwargs: P.kwargs) -> T: ...
```
This would mean that `target` must be both a `T` and `SupportsCopyFrom` simultaneously. Intersections I think are being considered more seriously than HKT, but they're also another quite substantial addition to the type system.

barneygale replied:
Thanks for the analysis! Feels like we're pushing the current type system a bit far, so I'm happy with `Any`. I could log an issue in pathlib-abc and link back to this thread.

I'm also happy to change how `copy()` and `_copy_from()` interact if there's something better we can do.

TeamSpen210 replied:
I think I have a solution, though it's awkward. First change is to pass `*args` too, meaning `ParamSpec` works. Second is to require `_copy_to()` to redundantly return `self`. Then we could do the following:
```py
class _HasCopyTo[**Args, R](Protocol):
    def _copy_to(
        self, path: ReadablePath, /, 
        *args: Args.args, **kwargs: Args.kwargs) -> R: ...

class ReadablePath:
    def copy[**Args, R](
        self, path: _HasCopyTo[Args, R], /, 
        *args: Args.args, **kwargs: Args.kwargs,
    ) -> R:
        res = path._copy_to(self, *args, **kwargs)
        assert path is res, "Must return self"
        return res

class WritablePath(ReadablePath):
    @abstractmethod
    def _copy_to(self, path: ReadablePath, /, *args: Any, **kwargs: Any) -> Self: ...

class Path(WritablePath):
    def _copy_to(self, path: ReadablePath, /, some_option: bool) -> Self:
        do_copy(self, path)
        return self
```
The `Any` in `WritablePath` is required to allow subclasses to require any arguments, but it also (correctly?) means if you call `copy()` with something that's just `WritablePath` it'll accept anything.

barneygale replied:
That LGTM - hopefully it can be expressed in 3.9 syntax?

ap-- replied:
It does! I'll make the change:

https://mypy-play.net/?mypy=latest&python=3.11&gist=8c300324c700eea4ffa7b7776a460f6e

<details>
<summary> code </summary>

```python
from typing import Literal
from typing import Protocol
from typing import Generic
from typing import TypeVar
from typing import ParamSpec
from typing import Callable
from typing import Self
from typing import ClassVar
from typing import Any

R = TypeVar("R", bound="_WritablePath", covariant=True)
P = ParamSpec("P")

class _HasCopyFrom(Protocol[R, P]):
    def _copy_from(self, source: _ReadablePath, /, *args: P.args, **kwargs: P.kwargs) -> R:
        ...

class _ReadablePath:
    def copy(self, path: _HasCopyFrom[R, P], /, *args: P.args, **kwargs: P.kwargs) -> R:
        return path._copy_from(self, *args, **kwargs)

class _WritablePath:
    def _copy_from(self, path: _ReadablePath, /, *args: Any, **kwargs: Any) -> Self: ...

class MyPathR(_ReadablePath):
    pass

class MyPathW0(_WritablePath):
    def _copy_from(self, path: _ReadablePath, /, *, some_option: bool = True) -> Self:
        return self

class MyPathW1(_WritablePath):
    def _copy_from(self, path: _ReadablePath, /, *, other_option: int = 3) -> Self:
        return self

r = MyPathR()       
w0 = MyPathW0()
w1 = MyPathW1()
 
# correct       
reveal_type(r.copy(w0))
r.copy(w0, some_option=False)
reveal_type(r.copy(w1))
r.copy(w1, other_option=4)

# error
r.copy(w0, dont_exist=2)
r.copy(w1, text="str")
r.copy(w0, some_option=1.0)

```

</details>

ap-- replied:
Hmmm, I tried implementing the changes, and they don't fully work. While they solve the copy (`*args` and) `**kwargs` issue, they cause type checking issues in for example `copy_into` because mypy can't infer the actual type of path, but only that path has the structural type `_HasCopyFrom`. This is a limitation of how the protocol is expressed (see here: https://mypy-play.net/?mypy=latest&python=3.12&gist=c744090b69aad82d75c14e6ab7dedb4c ). So it seems intersection types would be required to do this correctly.


Also on a side note: I now wonder if `follow_symlinks` should be an explicit keyword argument in `copy` and `copy_into` since it's mandatory in `_WritablePath._copy_from`'s default implementation.

Additionally, forcing `_copy_from` to return `Self` would interfere with #131636

TeamSpen210 replied:
[Fixed](https://mypy-play.net/?mypy=latest&python=3.12&gist=a19d8447e8e3b7d6317c8e5c734657ea) the issues - you need to be returning `R` from `_copy_from`, and `_HasCopyFrom.__truediv__`/`joinpath` needs to return `_HasCopyFrom`, not just `R`. That was causing it to discard the paramspec and specific return value. For #131636, you can still do `-> Generator[tuple[...], None, Self]`.

ap-- replied:
Yeah, but the problem imo is that both `joinpath` and `__truediv__` should actually return the same type. The workaround is nice, but to make it work, one needs to be typed as returning `R` because it has to return the WritablePath instance and the other as the `HasCopyFrom` protocol because it's passed again into .copy ...

So in the end you have to adjust the protocol to the exact implementation of copy and copy_into which is unfortunate. If both in copy and copy_into joinpath would be used once as the return type and once to pass as the target arg to copy, the typing workaround would not work.

I'm not sure if the benefit of being able to type check the copy kwargs justifies the introduction of this trick.

barneygale replied:
Agree with @ap--. We could always revisit this later. Thank you both for your work on this, it's been really valuable.

barneygale commented on Lib/pathlib/types.py:
```diff
@@ -14,10 +14,15 @@
+    runtime_checkable,
+)
 
+_WP = TypeVar("_WP", bound="_WritablePath")
```
Would `_CopyTarget` be a better name, now that this is only used from `copy()` and `copy_into()`?

(Not a leading question - I have very little exposure to type vars)

ap-- replied:
I renamed to `_WritablePathT`

Newer syntax would allow us to get rid of the TypeVar import and write:

```
   def copy[T: _WritablePath](self, target: T, **kwargs: Any) -> T:
```

But it'll increase friction with backporting

ap-- committed:
Switch walk to Iterator[...]

ap-- committed:
Use abstract base classes from collections.abc

ap-- committed:
Use _PathParser in _explode_path

ap-- committed:
Rename TypeVar to _WritablePathT

barneygale commented on Lib/pathlib/types.py:
```diff
@@ -11,18 +11,25 @@
 
-def _explode_path(path, split):
+
+def _explode_path(path: str, parser: "_PathParser") -> tuple[str, list[str]]:
```
Could move this function below `_PathParser` to avoid the forward ref?

ap-- commented on Lib/pathlib/types.py:
```diff
@@ -397,7 +420,7 @@ def write_text(self, data, encoding=None, errors=None, newline=None):
             return f.write(data)
 
-    def _copy_from(self, source, follow_symlinks=True):
+    def _copy_from(self, source: _ReadablePath, follow_symlinks: bool = True) -> None:
```
@barneygale I noticed `_copy_from` is currently less strict with additional kwargs compared to copy() and copy_into(). Shouldn't these be keyword only args?

```suggestion
    def _copy_from(self, source: _ReadablePath, *, follow_symlinks: bool = True) -> None:
```

barneygale replied:
Good spot, yes you're right!

ap-- committed:
Move _explode_path

ap-- committed:
Update Lib/pathlib/types.py

ap-- committed:
Merge remote-tracking branch 'origin/main' into pathlib-types-add-typehints

ap-- committed:
pathlib.types: don't rely on typing.Optional

ap-- committed:
avoid typing imports that are not needed at runtime

ap-- committed:
add pathlib to Misc/mypy tests

ap-- committed:
pathlib: fix most typing issues

ap-- committed:
prevent missing glob typing to bleed into _pyrepl mypy tests

ap-- committed:
workaround  crashing mypy
